rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,501 @@
+  """A `DataHandler` that is compatible with `ClusterCoordinator`."""
+
+  def _configure_dataset_and_inferred_steps(self, strategy, x, steps_per_epoch,
+                                            class_weight):
+    self._dataset = self._model._cluster_coordinator.create_per_worker_dataset(x)
+    if steps_per_epoch is None:
+      raise RuntimeError(
+          "Steps per epoch must be specified with `ParameterServerStrategy`.")
+    self._inferred_steps = steps_per_epoch
+
+  def sync(self):
+    self._model._cluster_coordinator.join()
+
+  def resolve_logs(self, logs):
+    return logs.fetch()
+```
+
+
+And, in the existing `DataHandler`,
+
+
+```
+class DataHandler(object):
+
+  def _configure_dataset_and_inferred_steps(self, strategy, x, steps_per_epoch,
+                                            class_weight):
+    # Does what it does currently
+
+  def sync(self):
+    context.async_wait()
+
+  @contextlib.contextmanager
+  def catch_stop_iteration(self):
+    """Catches errors when an iterator runs out of data."""
+    try:
+      yield
+      self.sync()
+    except (StopIteration, errors.OutOfRangeError):
+      # stop the iteration
+
+  def resolve_logs(self, logs):
+    return logs
+```
+
+
+A new `DataAdapter` is also needed to make sure the training API knows a dataset factory is a supported path:
+
+
+```
+class DataFactoryAdapter(DataAdapter):
+  """Adapter that handles dataset functions."""
+
+  @staticmethod
+  def can_handle(x, y=None):
+    if isinstance(x, DatasetFactory):
+      assert y is None
+      return True
+    return False
+```
+
+
+
+#### Callbacks
+
+With `ParameterServerStrategy`, the return value of `Model.train_function` is a dict `RemoteValue`s. This dict is passed as the `logs` argument to the `CallbackList` object. To obtain the `NumPy` value of each `RemoteValue`, one can do:
+
+
+```
+def to_numpy(logs):
+  cluster_coordinator.join()  # Sync the workers.
+  return {k: v.fetch() for k, v in logs.items()}  # Return the NumPy results
+```
+
+
+This logic can be handled by the `CallbackList` object, which already handles converting `tf.Tensor` logs to `NumPy`. However, obtaining these logs requires us to sync the workers, which will result in a slowdown if done every batch. Currently we only plan to sync the workers once every epoch at the end of the epoch. 
+
+
+##### Epoch-level callbacks
+
+Since the workers will sync every epoch anyway, fetching the remote values incurs minimal additional overhead and so epoch-level `Callback`s can be easily supported.
+
+
+##### Batch-level callbacks
+
+Some users may want to use batch-level `Callback`s. When users use `steps_per_execution=N`, the `Callback`s will only execute every `N` steps, and so batch-level callbacks might not be prohibitively slow for large `N`. However, in most cases, batch-level callbacks will cause a significant slowdown and are likely to be added only in error. We have a few options for handling batch-level Callbacks.
```

Comment:
I disagree, batch-level callbacks are useful and important. We just need to make sure that they're called at an appropriate frequency (especially by default)